### ISSUE TITLE
TChain issue in PyROOT with conda

### ISSUE THREAD

**Issue by Allen319** [root-project/root#6467]
This issue only occurs when the conda environment is activated.
I followed the default install method with `conda create -c conda-forge --name <my-environment> root`

I want to use a wrapper class for TChain called [Ntuple](https://github.com/Allen319/Ntuple/blob/master/python/Ntuple.py)
https://github.com/Allen319/Ntuple/blob/master/python/Ntuple.py
But I cannot initialize this class with conda environment (If I use ROOT installed from source codes, it works well).
The error message:
```
Traceback (most recent call last):
  File "./example_ntuple.py", line 9, in <module>
    n = Ntuple("DataTree")
  File "/Users/allen/Documents/GitHub/tuple/python/Ntuple.py", line 29, in __init__
    TChain.__init__(self, name)
TypeError: none of the 2 overloaded methods succeeded. Full details:
  TChain::TChain() =>
    TypeError: takes at most 0 arguments (1 given)
  no constructor available for 'TChain'
```
It is strange that there is no other overloaded method with 2 args

**Comment by etejedor** [comment]
Thank you for the report, I tried to import your `Ntuple` module and what I get with master and 6.22 patches (compiled from sources) is this:
```
>>> import Ntuple
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/etejedor/Ntuple.py", line 4, in <module>
    class Ntuple(ROOT.TChain):
KeyError: 'fProofChain'
```
So there is an error even before constructing the TChain, during inheritance. This error is due to the fact that `fProofChain` is a protected data member of `TChain`, so the issue is the same as for:
https://sft.its.cern.ch/jira/browse/ROOT-11025
https://sft.its.cern.ch/jira/browse/ROOT-10826

On the other hand, if I try both conda ROOT 6.22/02 and master, I indeed get the error you reported when constructing the TChain. This is strange, since the error in the inheritance should definitely happen before. @chrisburr can you think of why there is this difference (e.g. a patch that is applied for the conda build)?

**Comment by chrisburr** [comment]
There definitely shouldn't be anything weird about the conda build. I'll try looking closer and get back to you.

**Comment by chrisburr** [comment]
This is all that changed in the conda-recipe as part of updating to 6.22.0:

https://github.com/conda-forge/root-feedstock/pull/89/files

There doesn't seem to be anything suspicious

**Comment by Allen319** [comment]
Thank you both for the check.
Indeed, if you are quite certain with update changes of conda, maybe it has nothing to do with conda environment. 
For me, It works with root 6.18/04 built from source code.
So it's probably a version-related issue and not a conda one.

**Comment by etejedor** [comment]
I believe it depends indeed on the Python version. I am able to reproduce the inheritance error (`KeyError: 'fProofChain'`) with Python2.7 and 3.6.8, but not with 3.8.2.

**Comment by chrisburr** [comment]
I've just had a look with creating environments and indeed, different Python versions behave very differently and the segfaults go away if I use `python` instead of `ipython`.

### Examples

Each was created with:

```bash
mamba create --name test root python=X.Y
```

### Python 2.7

```python
Python 2.7.18 |Anaconda, Inc.| (default, Apr 23 2020, 22:17:58)
Type "copyright", "credits" or "license" for more information.

IPython 5.8.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import ROOT
   ...: import numpy as np
   ...:
   ...: class Ntuple(ROOT.TChain):
   ...:     def __init__(self, name):
   ...:         pass
   ...:
   ...: Ntuple("DataTree")
   ...:
input_line_38:9:43: error: base class '::TChain' has private copy constructor
  Dispatcher1(const Dispatcher1& other) : TChain(other), m_self(other.m_self, this) {}
                                          ^
/home/cburr/miniconda3/envs/test/include/TChain.h:49:4: note: declared private here
   TChain(const TChain&);            // not implemented
   ^
/home/cburr/miniconda3/envs/test/bin/ipython:4: RuntimeWarning: no python-side overrides supported
  import re
Out[1]: <cppyy.gbl.TChain object at 0x(nil)>
```

### Python 3.6

```python
Python 3.6.11 | packaged by conda-forge | (default, Aug  5 2020, 20:06:33)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.16.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import ROOT
   ...: import numpy as np
   ...:
   ...: class Ntuple(ROOT.TChain):
   ...:     def __init__(self, name):
   ...:         pass
   ...:
   ...: Ntuple("DataTree")
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-1-cdc669878b88> in <module>
      2 import numpy as np
      3
----> 4 class Ntuple(ROOT.TChain):
      5     def __init__(self, name):
      6         pass

KeyError: 'fProofChain'
```

### Python 3.7

```python
Python 3.7.8 | packaged by conda-forge | (default, Jul 31 2020, 02:23:50)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.18.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import ROOT
   ...: import numpy as np
   ...:
   ...: class Ntuple(ROOT.TChain):
   ...:     def __init__(self, name):
   ...:         pass
   ...:
   ...: Ntuple("DataTree")
Out[1]:  *** Break *** segmentation violation
 *** Break *** segmentation violation
```

### Python 3.8

```python
Python 3.8.5 | packaged by conda-forge | (default, Sep 24 2020, 16:55:52)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.18.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import ROOT
   ...: import numpy as np
   ...:
   ...: class Ntuple(ROOT.TChain):
   ...:     def __init__(self, name):
   ...:         pass
   ...:
   ...: Ntuple("DataTree")
Out[1]:  *** Break *** segmentation violation
 *** Break *** segmentation violation
```

**Comment by etejedor** [comment]
Thanks @chrisburr , these are all with 6.22/02, right? In master there are a few fixes for the inheritance, so I would try with it instead.

From the tests I have done so far (master):
- Python <= 3.6 : error when defining the derived Python class (`KeyError: 'fProofChain'`)
- Python >= 3.7 : error during the construction of the derived class object (`no constructor available for 'TChain'`)

I didn't try python/ipython.

**Comment by chrisburr** [comment]
Python 3 was 6.22.2, Python 2.7 used 6.22.0. (Python 2.7 isn't supported in conda-forge any more and I rarely make the effort to rebuild the py2k branch.)

I've stopped looking now as I think you're in a much better position than me to make sense of what the issue is but ping me if I can be useful.

**Comment by etejedor** [comment]
Yes thanks, I'll report when I have news.

**Comment by etejedor** [comment]
The second kind of error (`no constructor available for 'TChain'`) happens when invoking the constructor of TChain from `__init__`, in particular an overload with default arguments. The issue is that a match does not happen with the `(const char *name, const char *title="")` overload when just providing `name`. This needs to be fixed.

@Allen319 , as a temporary workaround in case you need it, you can do `n = Ntuple("DataTree", "")` instead of `n = Ntuple("DataTree")` and that should prevent the constructor error.

**Comment by etejedor** [comment]
Fix backported to 6.22 here: https://github.com/root-project/root/pull/6598
